xposed-bridge/src/main/java/de/robv/android/xposed/XSharedPreferences.java

```diff
@@ -197,26 +172,52 @@ public XSharedPreferences(String packageName, String prefFileName) {
     }
 
     private void tryRegisterWatcher() {
-        if (!mWatcherEnabled) {
+        if (mWatchKey != null && mWatchKey.isValid()) {
             return;
         }
-        Path path = mFile.toPath();
-        if (sInstances.containsKey(path)) {
-            return;
+
+        synchronized (sWatcherKeyInstances) {
+            Path path = mFile.toPath();
+            try {
+                if (sWatcher == null) {
+                    sWatcher = new File(XposedInit.prefsBasePath).toPath().getFileSystem().newWatchService();
+                    if (BuildConfig.DEBUG) Log.d(TAG, "Created WatchService instance");
+                }
+                mWatchKey = path.getParent().register(sWatcher, StandardWatchEventKinds.ENTRY_CREATE,
+                        StandardWatchEventKinds.ENTRY_MODIFY, StandardWatchEventKinds.ENTRY_DELETE);
+                sWatcherKeyInstances.put(mWatchKey, new PrefsData(this));
+                if (sWatcherDaemon == null || !sWatcherDaemon.isAlive()) {
+                    initWatcherDaemon();
+                }
+                if (BuildConfig.DEBUG) Log.d(TAG, "tryRegisterWatcher: registered file watcher for " + path);
+            } catch (AccessDeniedException accDeniedEx) {
+                if (BuildConfig.DEBUG) Log.e(TAG, "tryRegisterWatcher: access denied to " + path);
+            } catch (Exception e) {
+                Log.e(TAG, "tryRegisterWatcher: failed to register file watcher", e);
+            }
         }
-        try {
-            path.getParent().register(getWatcher(), StandardWatchEventKinds.ENTRY_CREATE, StandardWatchEventKinds.ENTRY_MODIFY, StandardWatchEventKinds.ENTRY_DELETE);
-            sInstances.put(path, new PrefsData(this));
-            if (BuildConfig.DEBUG) Log.d(TAG, "tryRegisterWatcher: registered file watcher for " + path);
-        } catch (AccessDeniedException accDeniedEx) {
-            if (BuildConfig.DEBUG) Log.d(TAG, "tryRegisterWatcher: access denied to " + path);
-        } catch (Exception e) {
-            Log.d(TAG, "tryRegisterWatcher: failed to register file watcher", e);
+    }
+
+    private void tryUnregisterWatcher() {
+        synchronized (sWatcherKeyInstances) {
+            if (sWatcherKeyInstances.containsKey(mWatchKey)) {
```

Comment:
Use remove() directly here and use it's return value to check if element existed before.